Add more context to `async fn` trait error (#66526)
rust-lang/rust: src/librustc_error_codes/error_codes/E0706.md

>  `async fn`s are not yet supported in Rust.

Reviewer: When I read this I was like "but it is?" -- can we clarify that this is about traits early on?

Erroneous code example:

```compile_fail,edition2018
trait T {
    // Neither case is currently supported.
    async fn foo() {}
    async fn bar(&self) {}
}
```

> `async fn`s normally return an `impl Future`, making the following two examples equivalent:

Reviewer: ```suggestion
`async fn`s return an `impl Future`, making the following two examples equivalent:
```

```edition2018,ignore (example-of-desugaring-equivalence)
async fn foo() -> User {
    unimplemented!()
}
// The async fn above gets desugared as follows:
fn foo(&self) -> impl Future<Output = User> + '_ {
    unimplemented!()
}
```

But when it comes to supporting this in traits, there are [a few implementation

> issues][async-is-hard], one of which is that returning `impl Trait` in traits is not supported,

Reviewer: Long sentence, split up?

as it would require [Generic Associated Types] to be supported:

```edition2018,ignore (example-of-desugaring-equivalence)
impl MyDatabase {
    async fn get_user(&self) -> User {
        unimplemented!()
    }
}

impl MyDatabase {
    fn get_user(&self) -> impl Future<Output = User> + '_ {
        unimplemented!()
    }
}
```

> Until these issues are resolved, you can use the [`async-trait` crate], which allows you to use

Reviewer: ```suggestion
Until these issues are resolved, you can use the [`async-trait` crate],
allowing you to use `async fn` in traits by desugaring to "boxed futures"
```

> this feature by sidesteping the language feature issue by desugaring to "boxed futures"

Reviewer: ```suggestion
```

(`Pin<Box<dyn Future + Send + 'async>>`):

> ```edition2018,ignore (example-of-desugaring-equivalence)

Reviewer: I would prefer to just link to the crate and let that explain what is happening in case e.g. the desugaring changes or whatnot.
